python/pyarrow/tests/parquet/test_dataset.py

```diff
@@ -1239,6 +1241,71 @@ def test_dataset_partitioning(tempdir):
     assert result.column_names == ["a", "year", "month", "day"]
 
 
+@pytest.mark.parametrize(
+    "flavor, expected_defined_partition, expected_undefined_partition",
+    [
+        (ds.HivePartitioning, (r"foo=A/bar=ant%20bee", ""), ("", "")),
+        (ds.DirectoryPartitioning, (r"A/ant bee", ""), ("", "")),
+        (ds.FilenamePartitioning, ("", r"A_ant bee_"), ("", "_")),
+    ],
+)
+def test_dataset_partitioning_format(
+    flavor: ds.Partitioning,
+    expected_defined_partition: tuple,
+    expected_undefined_partition: tuple,
+):
+
+    partitioning_schema = pa.schema([("foo", pa.string()), ("bar", pa.string())])
+
+    partitioning = flavor(schema=partitioning_schema)
+
+    # test forward transformation (format)
+    assert (
+        partitioning.format((pc.field("bar") == "ant bee") & (pc.field("foo") == "A"))
+        == expected_defined_partition
+    )
+
+    # test backward transformation (parse)
+    assert partitioning.parse("/".join(expected_defined_partition)).equals(
+        (pc.field("foo") == "A") & (pc.field("bar") == "ant bee")
+    )
+
+    # test complex expression can still be parsed into useful directory/path
+    assert (
+        partitioning.format(
+            ((pc.field("bar") == "ant bee") & (pc.field("foo") == "A"))
+            & ((pc.field("bar") == "ant bee") & (pc.field("foo") == "A"))
+        )
+        == expected_defined_partition
+    )
+
+    # test a different complex expression cannot be parsed into directory/path
+    # and just returns the same value as if no filter were applied.
+    assert (
+        partitioning.format(
+            ((pc.field("bar") == "ant bee") & (pc.field("foo") == "A"))
+            | ((pc.field("bar") == "ant bee") & (pc.field("foo") == "A"))
+        )
+        == expected_undefined_partition
+    )
+
+    if flavor != ds.HivePartitioning:
+        # Raises error upon filtering for lower level partition without filtering for
+        # higher level partition
+        with pytest.raises(ArrowInvalid) as raised:
+            partitioning.format(((pc.field("bar") == "ant bee")))
+        assert raised.type is ArrowInvalid
+        assert raised.value.args == (
+            "No partition key for foo but a key was provided subsequently for bar.",
```

Comment:
Can this be rewritten as this? It works for me locally:

```suggestion
        with pytest.raises(ArrowInvalid, match="No partition key for foo but a key was provided subsequently for bar"):
            partitioning.format(((pc.field("bar") == "ant bee")))
```

I actually found this because the tests were failing for me locally because apparently extra debug info ends up in the messages so the match wasn't exact.